### PR TITLE
Format insensitive cards for loader & Blank Delim for sideboard

### LINKED ISSUE #2701: Suggestion: remove case-sensitivity in "Import from Clipboard" in Deck Editor
OS and version irrelevant.

__________

The idea should be obvious. Case sensitivity doesn't help at all, and creates a lot of unnecessary retyping. Also would fix the "AEther" "Aether" issue.

### PR DESCRIPTION
**Pull request by ZeldaZach**
## Related Ticket(s)
- Fixes #2701 

## Short roundup of the initial problem
Cards input to the deck editor weren't validated correctly and, if not typed _exactly_ as they are defined in an XML file, would be added to the "unknown" section.

The sideboard wouldn't begin if there was a line break in a file and insisted on a "Sideboard" word or "SB:" in front of cards.

## What will change with this Pull Request?
This PR adds a validation against the installed databases to find which card the inputs match up to and gives them the correct name attribute before loading them to the deck editor, effectively allowing case insensitivity and grammatical insensitivity. 

This PR also addresses an issue when importing decks from MTGGoldFish where they would have the sideboard split up with one blank line. This is now an added feature to check for a blank line and then start the sideboard.

~~It is possible (in the future?) that we should re-address how the loader works and make it "smarter" to check if there is the word "Sideboard" in the file, and if it is there then treat blank lines like nothing.~~

I added some new parts to the parser to make it "smarter," per say. It will now cut out all blank lines at the start of the paste and see how many blank lines there are. If there's only 1, assume it's the sideboard split point. If not, just assume the old logic was correct.

### CODE REVIEW

File: common/decklist.cpp [hunk 1]
```diff
@@ -4,6 +4,8 @@
 #include <QCryptographicHash>
 #include <QDebug>
 #include "decklist.h"
+#include "../cockatrice/src/carddatabase.h"
```
Comment (ctrlaltca) [inline common/decklist.cpp]: This can't work, since decklist.cpp is a common file used by both cockatrice and servatrice.
I recently had to face the same problem in this code to fix token loading (see #2670, commit 36f6907fa36b60bcc03e93ca08fd7b3e2da56533).
The idea is that since cockatrice's DeckLoader is a subclass of the common class DeckList, you can create a new virtual method in DeckList and reimplement it in DeckLoader in order to add the needed functionality.
In my example i added the getCardZoneFromName method to get the correct zone for a card from the db.
In your code, you could do the same to add a method that renames the card with its correct name.

Comment (ZeldaZach) [inline common/decklist.cpp]: Well that stinks, I'll try to figure it out.

File: common/decklist.h [hunk 2]
```diff
@@ -125,6 +125,7 @@ class DeckList : public QObject {
     InnerDecklistNode *getZoneObjFromName(const QString  zoneName);
 protected:
     virtual QString getCardZoneFromName(QString /* cardName */, QString currentZoneName) { return currentZoneName; };
+    virtual QString getCompleteCardName(QString cardName) { return cardName; };
```
Comment (ZeldaZach) [inline common/decklist.h]: Oh go away, I used it.

**Comment by Nightfirecat** [comment]
Side thought: If the codebase has a good place to implement tests, decklist parsing seems like a very good candidate for some unit tests, particularly with these new changes.

**Comment by tooomm** [comment]
@MarkyMarkMcDonald added some tests recently. Want to have a look here as well?

**Comment by ZeldaZach** [comment]
The newest change I just pushed addresses in the inputs of all separate types. Will create testing soon. I added significant comments to help explain what's going on.

1) Trim all blank lines from start/end of input
    - ["", "", "c", "", "d", ""] => ["c", "", "d"]
2) Trim excessive blanks from interior
    - ["c", "", "", "", "d", "e"] => ["c", "", "d", "e"]
3) Check if there is a _single_ blank line in the file, and if so use it as the split point for sideboard
4) If multiple blank lines found, look for "Sideboard" to appear for split point.

**Comment by tooomm** [comment]
>If multiple blank lines found, look for "Sideboard" to appear for split point.

Your comment in the code is the other way around, check for "Sideboard" word, then blank lines.

You don't check for "SB:" anymore? Should we scrap that?

File: cockatrice/src/deck_loader.cpp [hunk 3]
```diff
@@ -200,9 +200,19 @@ void DeckLoader::saveToStream_DeckZoneCards(QTextStream &out, const InnerDecklis
...
 {
     CardInfo *card = db->getCard(cardName);
-    if(card && card->getIsToken())
+    if (card && card->getIsToken())
         return DECK_ZONE_TOKENS;
 
     return currentZoneName;
 }
 
+QString DeckLoader::getCompleteCardName(QString cardName)
```
Comment (backendr) [inline cockatrice/src/deck_loader.cpp]: Isnt this method name misleading? Why do you mutate the original?
`QString DeckLoader::getCompleteCardName(const QString cardName) const` may be a better way to write a getter; that or change the name?

Comment (ZeldaZach) [inline cockatrice/src/deck_loader.cpp]: I have to do this in order to access the database from the common folder. It's based on [Fabio's Comment](https://github.com/Cockatrice/Cockatrice/pull/2706#discussion_r115417049)... Please read all parts of the discussion

**Comment by ZeldaZach** [comment]
>You don't check for "SB:" anymore? Should we scrap that?
@tooomm We still check for this, just I explicitly wrote down in my comment what I changed. Feel free to read the actual code

File: cockatrice/src/deck_loader.cpp (hunk 3 above)
Comment (backendr) [inline cockatrice/src/deck_loader.cpp]: This doesn't explain why you wrote a mutating method that implies its a getter. Compared to `getCardZoneFromName` from the same header file which does not mutate parameters.

Comment (ctrlaltca) [inline cockatrice/src/deck_loader.cpp]: Well, @backendr is right, you should `return temp->getName();` if the card is found or `return cardName` if it-s not found, but not modify the contents of cardName itself.

Comment (ZeldaZach) [inline cockatrice/src/deck_loader.cpp]: Sounds good

Comment (backendr) [inline cockatrice/src/deck_loader.cpp]: To further enforce in the method signature your intentions, its best to change to `QString DeckLoader::getCompleteCardName(const QString cardName) const`

Comment (ZeldaZach) [inline cockatrice/src/deck_loader.cpp]: Sure @backendr. Changed

File: cockatrice/src/deck_loader.cpp [hunk 4]
```diff
@@ -200,9 +200,19 @@ void DeckLoader::saveToStream_DeckZoneCards(QTextStream &out, const InnerDecklis
...
 {
     CardInfo *card = db->getCard(cardName);
-    if(card && card->getIsToken())
+    if (card && card->getIsToken())
         return DECK_ZONE_TOKENS;
 
     return currentZoneName;
 }
 
+QString DeckLoader::getCompleteCardName(QString cardName) const
```
Comment (backendr) [inline cockatrice/src/deck_loader.cpp]: The passed in `QString cardName` can be made `const QString cardName`. This ensures that the passed in param is not mutated. 
The trailing method `const` only states this method will not mutate class state.

File: common/decklist.cpp [hunk 5]
```diff
@@ -466,29 +468,68 @@ bool DeckList::saveToFile_Native(QIODevice *device)
 bool DeckList::loadFromStream_Plain(QTextStream &in)
 {
     cleanList();
+    bool inSideboard = false, titleFound = false, isSideboard;
+    int okRows = 0, blankLines = 0, linesToSkip = 0;
```
Comment (backendr) [inline common/decklist.cpp]: `linesToSkip` is never used

File: common/decklist.cpp [hunk 6]
```diff
@@ -466,29 +468,74 @@ bool DeckList::saveToFile_Native(QIODevice *device)
 bool DeckList::loadFromStream_Plain(QTextStream &in)
 {
     cleanList();
+    bool inSideboard = false, titleFound = false, isSideboard;
+    int okRows = 0, blankLines = 0;
```
Comment (backendr) [inline common/decklist.cpp]: Its best to keep the usage of variables as close to their usage as possible. 
The bools and `okRows` can be moved to line `517`, `blankLines` to `505`

**Comment by backendr** [comment]
Having an additional linebreak in either the MB or SB will result in a 75 card MB.

```
2 Aether Hub
2 Archangel Avacyn
2 Canyon Slough
4 Concealed Courtyard
1 Cut
4 Fatal Push
4 Gideon, Ally of Zendikar
4 Heart of Kiran
4 Inspiring Vantage
1 Mountain
4 Plains
4 Scrapheap Scrounger
2 Shambling Vent
4 Spire of Industry
2 Swamp
2 Thalia, Heretic Cathar
4 Thraben Inspector
4 Toolcraft Exemplar
4 Unlicensed Disintegration
2 Walking Ballista

1 Anguished Unmaking
1 Cut
1 Fumigate
1 Magma Spray
2 Nahiri, the Harbinger
2 Oath of Liliana
1 Painful Truths

2 Release the Gremlins
2 Sorin, Grim Nemesis
2 Transgress the Mind

```

File: common/decklist.cpp [hunk 7]
```diff
@@ -537,20 +585,25 @@ bool DeckList::loadFromStream_Plain(QTextStream &in)
...
         rx.setPattern("([^A-Z])\\s*&\\s*");
-        if (rx.indexIn(cardName) != -1) {
+        if (rx.indexIn(cardName) != -1)
             cardName.replace(rx, QString("%1 // ").arg(rx.cap(1)));
-        }
 
-        // Look for the correct card zone
-        QString zoneName = getCardZoneFromName(cardName, isSideboard ? DECK_ZONE_SIDE: DECK_ZONE_MAIN);
+        // We need to get the name of the card from the database,
+        // but we can't do that until we sanitize the name
```
Comment (backendr) [inline common/decklist.cpp]: when does sanitization happening? I cant see a usage of it in `getCompleteCardName` or its stack.

File: common/decklist.h [hunk 8]
```diff
@@ -125,6 +125,7 @@ class DeckList : public QObject {
     InnerDecklistNode *getZoneObjFromName(const QString  zoneName);
 protected:
     virtual QString getCardZoneFromName(QString /* cardName */, QString currentZoneName) { return currentZoneName; };
+    virtual QString getCompleteCardName(QString cardName) const { return cardName; };
```
Comment (backendr) [inline common/decklist.h]: `virtual QString getCompleteCardName(const QString cardName) const { return cardName; };`

**Comment by ZeldaZach** [comment]
>Having an additional linebreak in either the MB or SB will result in a 75 card MB.

Expected behavior. I'm trying to make the parser smarter in a sense as it didn't account for blank lines in the past for a single line, but if you have multiple blank lines it's not my fault and you'll have to make your deck list more concise. I have no way to appropriately parse that information if there are blank lines like that everywhere.

File: common/decklist.cpp (hunk 7 above)
Comment (ZeldaZach) [inline common/decklist.cpp]: Bad wording, fixed

File: common/decklist.h (hunk 8 above)
Comment (ZeldaZach) [inline common/decklist.h]: 👍

File: common/decklist.cpp (hunk 6 above)
Comment (ZeldaZach) [inline common/decklist.cpp]: I'm fine leaving it where it is

File: cockatrice/src/deck_loader.cpp (hunk 4 above)
Comment (ZeldaZach) [inline cockatrice/src/deck_loader.cpp]: 👍

**Comment by ZeldaZach** [comment]
So it seems like my changes fail travis due to the function `QVector::rbegin()` being introduced in QT 5.6 and we compile against 5.0.3 and 5.4 (if I recall). @ctrlaltca Any alternatives you can suggest?

**Comment by Daenyth** [comment]
We should generalize the Dockerfile for servatrice to be able to compile and package everything outputting to a volume, then it doesn't matter how crappy travis's build architecture is

**Comment by ctrlaltca** [comment]
> the function QVector::rbegin() being introduced in QT 5.6

No reverse iterators before 5.6, sorry. You'll have to modify line https://github.com/Cockatrice/Cockatrice/pull/2706/files#diff-166b29bd05b87f762761b55ce34db82dR481 to use a backwards `for` loop or something similar.

**Comment by ZeldaZach** [comment]
@ctrlaltca Is there a way to require a later QT? Can we somehow bundle it?

File: common/decklist.cpp (hunk 6 above)
Comment (backendr) [inline common/decklist.cpp]: Why?

**Comment by ctrlaltca** [comment]
Requiring 5.6 for the whole project would mean cutting out some linux distributions.
You could `#ifdef` the code to be used only if at least Qt 5.6 is found, but probably rewriting it is a easier solution.

**Comment by ZeldaZach** [comment]
I created some new tests for the loader. If anyone else has some suggestions please let me know!

File: common/decklist.cpp (hunk 6 above)
Comment (ctrlaltca) [inline common/decklist.cpp]: It's quite a minor detail, but declaring variables just before their use eases to read and understand the code.
Sometimes creating scopes and declaring variables inside the new scope helps even more, because you'll be sure that nothing outside the scope can touch and modify the variable.
Again, it's quite a minor thing and cockatrice has probably a lot of places where code could be made easier to understand.

File: common/decklist.cpp [hunk 9]
```diff
@@ -463,32 +465,84 @@ bool DeckList::saveToFile_Native(QIODevice *device)
...
+
+    // Lets convert input QTextStream => QVector
+    while (!in.atEnd())
+        inputs.push_back(in.readLine().simplified().toLower());
+
+    // "Trim" the inputs of all blank lines at the start & end of paste
+    removeEmptyLinesFromStart(inputs);
+    std::reverse(inputs.begin(), inputs.end()); // Reverse vector (to remove blanks from end)
+    removeEmptyLinesFromStart(inputs);
+    std::reverse(inputs.begin(), inputs.end()); // Vector now in correct order
```
Comment (ctrlaltca) [inline common/decklist.cpp]: Lines 482-489 and the `removeEmptyLinesFromStart` method looks over engineered to me.
Empty lines at start/end could just be dropped inside the first loop, and even multiple empty lines could be condensed in the same loop, and the remaining empty lines counted.
This would avoid the need of add all the lines them to the vector, reverse the vector twice, call the obscure std-iterators-with-callback-loop-whatever at line 471 and them loop the vector again at line 493, and again at line 513.
Anyway, as long as it works.. Yeah, well, that's just, like, my opinion, man (cit.)

File: common/decklist.cpp [hunk 10]
```diff
@@ -463,32 +465,84 @@ bool DeckList::saveToFile_Native(QIODevice *device)
...
+
+    // This will "trim" multiple blank lines in a row to just one blank
```
Comment (ctrlaltca) [inline common/decklist.cpp]: I usually see the english term "condense" for such operations, maybe it fits better?

File: common/decklist.cpp [hunk 11]
```diff
@@ -463,32 +465,84 @@ bool DeckList::saveToFile_Native(QIODevice *device)
...
+    while (true) {
+        auto i1 = std::adjacent_find(inputs.begin(), inputs.end());
+
+        if (i1 != inputs.end()) {
+            if (i1 != QString()) {
+                // So we have a duplicate row like (..., "1 card1", "3 card1", ...)
+                // To deal with this, we will combine these rows together to form (..., "4 card1", ...)
+                // And replace the newest row (aka we'd edit "3 card1" in our example) with the new value
+                int combinedTotal = (*i1).mid(0,1).toInt() + (*i1+1).mid(0,1).toInt();
+                inputs.replace(static_cast<int>(std::distance(inputs.begin(), i1+1)), QString::number(combinedTotal) + (*(i1+1)).mid(1));
```
Comment (ctrlaltca) [inline common/decklist.cpp]: Is there a good reason to do this? From the comment above this loop was meant to only delete double empty lines.
Also, what happens if my deck does't contain card numbers, or has more that 9 cards (that is, a number with more than a single digit) with the same name?
Example 1:
```
Kitten
Kitten
Kitten
Kitten
```

Example 2:
```
12 Kitten
13 Kitten
```

File: common/decklist.cpp [hunk 12]
```diff
@@ -463,32 +465,84 @@ bool DeckList::saveToFile_Native(QIODevice *device)
...
+    // Or if the word "Sideboard" appears, then we just skip this check
+    for (auto iter = inputs.begin(); iter != inputs.end(); iter++) {
+        QString line = *iter;
 
-    int okRows = 0;
-    bool titleFound = false;
-    while (!in.atEnd()) {
-        QString line = in.readLine().simplified();
+        if (line.isEmpty())
+            blankLines++;
```
Comment (ctrlaltca) [inline common/decklist.cpp]: Yet another loop :(

File: common/decklist.cpp (hunk 11 above)
Comment (ZeldaZach) [inline common/decklist.cpp]: I can revoke this part as I just added it for personal adds

File: common/decklist.cpp (hunk 6 above)
Comment (backendr) [inline common/decklist.cpp]: This is what McConnell says in his must-read book Code Complete 2:

Ideally, declare and define each variable close to where it’s used. A declaration establishes a variable’s type. A definition assigns the variable a specific value. In languages that support it, such as C++ and Java, variables should be declared and defined close to where they are first used.

He also recommends keeping variables alive for as short a time as possible, and to minimize scope.

**Comment by ZeldaZach** [comment]
Ok I made another set of changes. I hopefully increased the efficiency of the program. Personally, the way I see it, this piece of code is run every once in a while, one time, by a user. It doesn't have to be the _most_ efficient thing in the client.

File: common/decklist.cpp [hunk 13]
```diff
@@ -466,29 +468,74 @@ bool DeckList::saveToFile_Native(QIODevice *device)
...
     int okRows = 0;
-    bool titleFound = false;
-    while (!in.atEnd()) {
-        QString line = in.readLine().simplified();
+    for (auto iter = inputs.begin(); iter != inputs.end(); iter++)
+    {
+        QString line = *iter;
 
-        // skip comments
+        // This is a comment line, ignore it
```
Comment (backendr) [inline common/decklist.cpp]: Can be removed

File: cockatrice/src/chatview/chatview.cpp [hunk 14]
```diff
@@ -93,7 +93,7 @@ void ChatView::appendCardTag(QTextCursor &cursor, const QString &cardName)
     anchorFormat.setAnchor(true);
     anchorFormat.setAnchorHref("card://" + cardName);
     anchorFormat.setFontItalic(true);
-    
+
```
Comment (ctrlaltca) [inline cockatrice/src/chatview/chatview.cpp]: Unrelated change

File: common/decklist.cpp [hunk 15]
```diff
@@ -466,31 +468,76 @@ bool DeckList::saveToFile_Native(QIODevice *device)
...
+        QString line = in.readLine().simplified().toLower();
+
+        /*
+         * Removes all blank lines at start of inputs
+         * Ex: ("", "", "", "Card1", "Card2") => ("Card1", "Card2")
+         *
+         * This will also concise multiple blank lines in a row to just one blank
+         * Ex: ("Card1", "Card2", "", "", "", "Card3") => ("Card1", "Card2", "", "Card3")
+         */
+        if (line == QString()) {
```
Comment (ctrlaltca) [inline common/decklist.cpp]: `if(line.isEmpty()) {`

File: common/decklist.cpp [hunk 16]
```diff
@@ -466,31 +468,76 @@ bool DeckList::saveToFile_Native(QIODevice *device)
...
+        inputs.push_back(line);
+    }
+
+    /*
+     * Removes  blank line at end of inputs (if applicable)
+     * Ex: ("Card1", "Card2", "") => ("Card1", "Card2")
+     * NOTE: Any duplicates were taken care of above, so there can be
+     * at most one blank line at the very end
+     */
+    if (inputs.size() && inputs.last() == QString())
```
Comment (ctrlaltca) [inline common/decklist.cpp]: `.. && inputs.last().isEmpty()`

File: common/decklist.cpp [hunk 17]
```diff
@@ -466,31 +468,76 @@ bool DeckList::saveToFile_Native(QIODevice *device)
...
+        blankLines = 2;
 
+    bool inSideboard = false, titleFound = false, isSideboard;
     int okRows = 0;
-    bool titleFound = false;
-    while (!in.atEnd()) {
-        QString line = in.readLine().simplified();
+    for (auto iter = inputs.begin(); iter != inputs.end(); iter++)
+    {
+        QString line = *iter;
```
Comment (ctrlaltca) [inline common/decklist.cpp]: Alternative: `foreach(QString line, inputs) {`

File: common/decklist.cpp [hunk 18]
```diff
@@ -500,32 +547,36 @@ bool DeckList::loadFromStream_Plain(QTextStream &in)
             isSideboard = true;
```
Comment (ctrlaltca) [inline common/decklist.cpp]: 2 lines above:
The line content is now surely lowercase, so: `if(line.startsWith("sb:")) {`

File: common/decklist.cpp [hunk 19]
```diff
@@ -500,32 +547,36 @@ bool DeckList::loadFromStream_Plain(QTextStream &in)
...
         }
 
+        if (line.trimmed().isEmpty())
+            continue; // The line was "    " instead of "\n"
```
Comment (ctrlaltca) [inline common/decklist.cpp]: Can you explain this a little bit more? Is it a check for empty SB: lines like
```
SB:       
```
Isn't line supposed to have already be trimmed by `simplified()` above?

File: common/decklist.cpp [hunk 20]
```diff
@@ -500,32 +547,36 @@ bool DeckList::loadFromStream_Plain(QTextStream &in)
...
         line.remove(rx);
         rx.setPattern("\\(.*\\)");
         line.remove(rx);
-        //Filter out post card name editions
+
+        // Filter out post card name editions
         rx.setPattern("\\|.*$");
         line.remove(rx);
-        line = line.simplified();
+        line = line.simplified().toLower();
```
Comment (ctrlaltca) [inline common/decklist.cpp]: Isn't line supposed to have already be `simplified()` and `toLower()`-ed above?

File: cockatrice/src/chatview/chatview.cpp (hunk 14 above)
Comment (ZeldaZach) [inline cockatrice/src/chatview/chatview.cpp]: No idea how to undo this, it's just spacing anyway, doesn't rlly matter

File: common/decklist.cpp (hunk 19 above)
Comment (ZeldaZach) [inline common/decklist.cpp]: While testing, for some reason entries as "______" started appearing randomly, and this stops it. There were no underscores in the inputs, and it was interpreting extra blank lines as this format for some reason. This just prevents that issue.

File: common/decklist.cpp (hunk 20 above)
Comment (ZeldaZach) [inline common/decklist.cpp]: Yeah...

File: cockatrice/src/chatview/chatview.cpp (hunk 14 above)
Comment (ctrlaltca) [inline cockatrice/src/chatview/chatview.cpp]: If this were a big project with a dozen people working on it, such a change would be a problem:
 1. because it could introduce merge conflicts
 2. because the git log will store anyway the fact that you are the author that  last changed that line of code (think of git blame, etc..)

In our case I agree it's not important, you can leave it as it is.

File: common/decklist.cpp (hunk 19 above)
Comment (ctrlaltca) [inline common/decklist.cpp]: Ok, so it's a sanity check on the line contents before starting to parse card name and qty.

File: cockatrice/src/chatview/chatview.cpp (hunk 14 above)
Comment (ZeldaZach) [inline cockatrice/src/chatview/chatview.cpp]: Thanks for the tip!

File: common/decklist.cpp (hunk 18 above)
Comment (ZeldaZach) [inline common/decklist.cpp]: dealt with

File: common/decklist.cpp (hunk 19 above)
Comment (ZeldaZach) [inline common/decklist.cpp]: Yes

File: cockatrice/src/chatview/chatview.cpp (hunk 14 above)
Comment (Daenyth) [inline cockatrice/src/chatview/chatview.cpp]: Tip: When making changes, use `git add -p` to prepare your commit diff by diff. That way you can sculpt specific commits, and do the behavior changes in one, and formatting/refactoring in another

File: cockatrice/src/deck_loader.cpp [hunk 21]
```diff
@@ -200,9 +200,19 @@ void DeckLoader::saveToStream_DeckZoneCards(QTextStream &out, const InnerDecklis
 QString DeckLoader::getCardZoneFromName(QString cardName, QString currentZoneName)
 {
     CardInfo *card = db->getCard(cardName);
-    if(card && card->getIsToken())
+    if (card && card->getIsToken())
```
Comment (Daenyth) [inline cockatrice/src/deck_loader.cpp]: I really need to get clang-format set up. It's not hard to run, the hard part is just tuning its configuration flags to be as close as possible to current style, so as not to rewrite the whole codebase

File: common/decklist.cpp [hunk 22]
```diff
@@ -466,66 +466,111 @@ bool DeckList::saveToFile_Native(QIODevice *device)
 bool DeckList::loadFromStream_Plain(QTextStream &in)
 {
     cleanList();
+    QVector<QString> inputs; // QTextStream -> QVector
 
-    bool inSideboard = false, isSideboard = false;
+    bool priorEntryIsBlank = true, isAtBeginning = true;
```
Comment (Daenyth) [inline common/decklist.cpp]: For the future let's not do same-line declarations

File: common/decklist.cpp [hunk 23]
```diff
@@ -466,66 +466,111 @@ bool DeckList::saveToFile_Native(QIODevice *device)
 bool DeckList::loadFromStream_Plain(QTextStream &in)
```
Comment (Daenyth) [inline common/decklist.cpp]: This function is massive, it would be good to break it up into smaller units that are more easily composed and tested in isolation

File: tests/loading_from_clipboard/loading_from_clipboard_test.cpp [hunk 24]
```diff
@@ -39,28 +39,33 @@ struct DecklistBuilder {
...
         DeckList *deckList = fromClipboard(new QString("Sideboard"));
-        ASSERT_TRUE(deckList->getCardList().isEmpty()) << "Deck should be empty";
+        ASSERT_TRUE(deckList->getCardList().isEmpty());
     }
 
     TEST(LoadingFromClipboardTest, QuantityPrefixed) {
         QString *clipboard = new QString(
                 "1 Mountain\n"
-                "2x Island\n"
+                        "2x Island\n"
```
Comment (Daenyth) [inline tests/loading_from_clipboard/loading_from_clipboard_test.cpp]: Extra spacing

File: tests/loading_from_clipboard/loading_from_clipboard_test.cpp [hunk 25]
```diff
@@ -114,12 +123,142 @@ namespace {
         DecklistBuilder decklistBuilder = DecklistBuilder();
         deckList->forEachCard(decklistBuilder);
 
-        CardRows expectedMainboard = CardRows({{"CardThatDoesNotExistInCardsXml", 1}});
+        CardRows expectedMainboard = CardRows({
+                                                      {"cardthatdoesnotexistincardsxml", 1}
```
Comment (Daenyth) [inline tests/loading_from_clipboard/loading_from_clipboard_test.cpp]: what's with the spacing here?

File: common/decklist.cpp (hunk 23 above)
Comment (ZeldaZach) [inline common/decklist.cpp]: Future endeavors

File: cockatrice/src/deck_loader.cpp (hunk 21 above)
Comment (ZeldaZach) [inline cockatrice/src/deck_loader.cpp]: Yeah....

File: common/decklist.cpp (hunk 22 above)
Comment (ZeldaZach) [inline common/decklist.cpp]: Isn't it more efficient to do so?

File: tests/loading_from_clipboard/loading_from_clipboard_test.cpp (hunk 25 above)
Comment (ZeldaZach) [inline tests/loading_from_clipboard/loading_from_clipboard_test.cpp]: CLion did it and i couldn't change it no matter what I did... maybe you can help?

File: common/decklist.cpp (hunk 22 above)
Comment (backendr) [inline common/decklist.cpp]: Its not inefficient, though it can be less readable, as one line is doing more than one thing. You gain saving a line over readability. I prefer readability.

File: common/decklist.cpp (hunk 23 above)
Comment (backendr) [inline common/decklist.cpp]: Do you have a timescale on when? Its better to "eat the frog" now than procrastinate on it. Its too easy to get a PR merged and either not come back to it, or leave more work for someone else.

File: tests/loading_from_clipboard/loading_from_clipboard_test.cpp (hunk 25 above)
Comment (backendr) [inline tests/loading_from_clipboard/loading_from_clipboard_test.cpp]: Try editing the file in your terminal

File: common/decklist.cpp (hunk 23 above)
Comment (ZeldaZach) [inline common/decklist.cpp]: I don't plan on coming back to this personally, it'll be someone else small project

File: tests/loading_from_clipboard/loading_from_clipboard_test.cpp (hunk 25 above)
Comment (ZeldaZach) [inline tests/loading_from_clipboard/loading_from_clipboard_test.cpp]: That's only a temporary fix, there's probably a CLion configuration that I can edit just not sure which

File: common/decklist.cpp (hunk 22 above)
Comment (Daenyth) [inline common/decklist.cpp]: It's strictly a readability thing

**Comment by Daenyth** [comment]
Is this waiting on anything? Looks good to go

**Comment by ctrlaltca** [comment]
This is good to go, except for the extra spacing you noticed.

**Comment by tooomm** [comment]
And https://github.com/Cockatrice/Cockatrice/pull/2706#discussion_r116356241 maybe?
